spec/queries/customers_with_balance_query_spec.rb

```diff
@@ -2,97 +2,100 @@
 
 require 'spec_helper'
 
-describe CustomersWithBalance do
-  subject(:customers_with_balance) { described_class.new(Customer.where(id: customer)) }
+describe CustomersWithBalanceQuery do
+  subject(:result) { described_class.new(Customer.where(id: customers)).call }
 
-  describe '#query' do
-    let(:customer) { create(:customer) }
+  describe '#call' do
+    let(:customers) { create(:customer) }
     let(:total) { 200.00 }
     let(:order_total) { 100.00 }
-    let(:outstanding_balance) { instance_double(OutstandingBalance) }
+    let(:outstanding_balance) { instance_double(OutstandingBalanceQuery) }
 
-    it 'calls CustomersWithBalance#statement' do
-      allow(OutstandingBalance).to receive(:new).and_return(outstanding_balance)
+    it 'calls OutstandingBalanceQuery#statement' do
+      allow(OutstandingBalanceQuery).to receive(:new).and_return(outstanding_balance)
       expect(outstanding_balance).to receive(:statement)
 
-      customers_with_balance.query
+      result
     end
 
     describe 'arguments' do
       context 'with customers collection' do
+        let(:customers) { create_pair(:customer) }
+
         it 'returns balance' do
-          customers = create_pair(:customer)
-          query = described_class.new(Customer.where(id: customers)).query
-          expect(query.pluck(:id).sort).to eq([customers.first.id, customers.second.id].sort)
-          expect(query.map(&:balance_value)).to eq([0, 0])
+          expect(result.pluck(:id).sort).to eq([customers.first.id, customers.second.id].sort)
+          expect(result.map(&:balance_value)).to eq([0, 0])
         end
       end
 
       context 'with empty customers collection' do
+        let(:customers) { Customer.none }
+
         it 'returns empty customers collection' do
-          expect(described_class.new(Customer.none).query).to eq([])
+          expect(result).to eq([])
         end
       end
     end
 
     context 'when orders are in cart state' do
       before do
-        create(:order, customer:, total: order_total, payment_total: 0, state: 'cart')
-        create(:order, customer:, total: order_total, payment_total: 0, state: 'cart')
+        create(:order, customer: customers, total: order_total, payment_total: 0, state: 'cart')
+        create(:order, customer: customers, total: order_total, payment_total: 0, state: 'cart')
```

Comment:
At first the `let(:customers)` refactor looked like a good idea, but now we're using it in different contexts, and here it's confusing because the name is plural and there's only one. It affects a number of specs below, so I'm not sure it makes the code better.

I can't think of a better way to arrange it right now, so would suggest reverting that change, would that be ok?

Reply:
@dacook thank you for your feedback. While there may not be official documentation explicitly stating these conventions, adhering to such practices is widely recognized within the development community making the codebase more readable and maintainable.

Regarding the `call` method, its use aligns with common design patterns in Ruby on Rails, providing a clear interface for executing main functionalities. Running the query in the initializer might be possible, but it could potentially lead to unexpected behavior, and it may not align with the typical design patterns in Rails. So I wouldn't suggest running any logic in the initializer :)

As for the `customers` variable in specs, I updated it to not affect `order` creation in specs below. cc: @rioug